update: payment provider protocol for point-of-sale (POS) guide (#2810)
vtexdocs/dev-portal-content: docs/guides/Integration-Guides/payments-integration-guide/payments-integration-ppp-applied-to-pos.md

To integrate a POS payment solution, payment partners must meet the following requirements:

- Develop a payment connector using the [Payment Provider Protocol (PPP)](https://help.vtex.com/en/tutorial/payment-provider-protocol--RdsT2spdq80MMwwOeEq0m). To handle POS payments, the connector requires the following:
  - Include the supported payment methods (credit or debit card) in the [Manifest](https://developers.vtex.com/docs/api-reference/payment-provider-protocol#get-/manifest).
  - Use [callbacks](https://help.vtex.com/en/tutorial/payment-provider-protocol--RdsT2spdq80MMwwOeEq0m#payment-authorization) for asynchronous payments.
  - Work with [Payment Apps](https://developers.vtex.com/docs/guides/payments-integration-payment-app) to identify the POS and wait for its response.
  - The endpoint must respond to any call in less than 20 seconds.

> - For testing, use a VTEX store configured with the supported payment methods. VTEX provides this store.

Reviewer: ```suggestion
- For testing, use a VTEX store configured with the supported payment methods. VTEX can provide an account for this purpose.
```

- Optionally, create a Payment App to identify the POS. Install this app in the store as well.

> ⚠️ To develop a new payment connector, you must follow the **prerequisites determined by VTEX**. See the [Implementation prerequisites section of the Payment Provider Protocol article](https://help.vtex.com/en/tutorial/payment-provider-protocol--RdsT2spdq80MMwwOeEq0m#implementation-prerequisites).

This guide covers the full journey to integrate POS payments:

- [Payment connector prerequisites](#payment-connector-prerequisites): Required routes and manifest configuration.
- [How POS payments work](#how-pos-payments-work): End-to-end payment flow, from identifying the POS to confirming the result.
- [Implementing the connector](#implementing-the-connector): Create Payment parameters, asynchronous flow, timeouts, Payment Apps, and POS interaction.
- [Testing your connector](#testing-your-connector): Setting up a store to test the integration.
- [Homologation](#homologation): Validating and publishing your connector.

## Payment connector prerequisites

Complete the following steps to enable the connector to process payments in the physical world:

> 1. Route [GET List Payment Provider Manifest](https://developers.vtex.com/docs/api-reference/payment-provider-protocol#get-/manifest) (`/manifest`). The response body of the endpoint must contain the payment methods `Venda Direta Credito` and `Venda Direta Debito`. If a split of receivables is offered, indicate it in the `allowsSplit` field for each payment method. The following example shows a correctly configured response body, including the POS-specific `Venda Direta Debito` and `Venda Direta Credito` methods:

Reviewer: ```suggestion
1. Route [GET List Payment Provider Manifest](https://developers.vtex.com/docs/api-reference/payment-provider-protocol#get-/manifest) (`/manifest`).

The response body of the endpoint must contain the payment methods `Venda Direta Credito` and `Venda Direta Debito`. If a split of receivables is offered, indicate it in the `allowsSplit` field for each payment method. The following example shows a correctly configured response body, including the POS-specific `Venda Direta Debito` and `Venda Direta Credito` methods:
```

Reviewer: I recommend doing this for numbered items in this section. I think having a whole paragraph as a numbered item is not super scannable.

Author: Great! Done!

2. The VTEX Sales App makes an [Authorization](https://developers.vtex.com/docs/api-reference/payments-gateway-api#post-/api/pvt/transactions/-transactionId-/authorization-request) request to the VTEX Payment Gateway.
3. The Gateway makes a [Create Payment](https://developers.vtex.com/docs/api-reference/payment-provider-protocol#post-/payments) request to the connector defined in the store settings for the specific payment method used in the purchase (for example, `Venda Direta Credito`).
4. The Gateway receives the Create Payment response from the connector with the `undefined` status and the `appname` field of the `paymentAppData` parameter filled with the name of an app (for example, `vtex.challenge-terminal-connector-app`). This means the payment is not concluded and requires a challenge, which opens a [Payment App](https://developers.vtex.com/docs/guides/payments-integration-payment-app#understanding-the-payment-app-flow) responsible for identifying the POS.
5. The VTEX Sales App receives the response from the Gateway and opens the app to start the challenge.

> 6. The seller interacts with the app to identify the POS used in the payment. The app then sends the POS serial number to the connector and closes.

Reviewer: Maybe 'Sales person' is best here. 'Seller' is ambiguous because it is most often used in our docs as the selling entity (oposed to a marketplace).

Author: Good! Updated!

7. The connector updates the order in the payment processor with the POS serial number.
8. After the app closes, the VTEX Sales App triggers a callback to the Gateway, starting the sequence of requests to process the payment on the POS.

>    1. The Gateway makes a Create Payment request to the connector with the same payment ID used in step 3, waiting for an update on the payment status.

Reviewer: Isn't this 'sub-list' suposed to be lettered instead of numbered (e.g., 'a, b, c', etc)? I am not sure, but I'd recommend taking a look at our guidelines.

Author: Perfect! Done!

To make the flow resilient:

- **Keep responses under 20 seconds.** Every connector endpoint must respond within 20 seconds; otherwise, the Gateway treats the call as failed.
- **Set `secondsWaiting` and `delayToCancel` consistently.** `delayToCancel` should be greater than or equal to `secondsWaiting` so the Gateway does not cancel a payment that the app is still waiting for.
- **Return the latest status on repeated Create Payment calls.** Because the route is idempotent, respond with the current status rather than creating a new payment.
- **Handle POS-side failures at the processor.** Card declines, insufficient funds, and password retries are resolved between the payment processor and the POS. Report the final result to the Gateway through the callback flow with `approved` or `denied`.
- **Confirm cancellation on failure.** When a payment is canceled or denied, the VTEX Sales App returns the buyer to checkout so they can retry.

### Payment Apps and paymentAppData

> At VTEX, payments in physical stores follow an asynchronous flow that requires additional interactions with the user, also called challenges, before approval. To request this interaction, the connector returns the `paymentAppData` field in the Create Payment response. This field contains the name of the VTEX IO app (`appName`) presented on the VTEX Sales App and the data the app needs to work (`payload`).

Reviewer: ```suggestion
At VTEX, payments in physical stores follow an asynchronous flow that requires additional interactions with the user, also called challenges, before approval. To request this interaction, the connector returns the `paymentAppData` field in the Create Payment response. This field contains the name of the VTEX IO app (`appName`) presented on the VTEX Sales App and the data the app needs to function (`payload`).
```
